### PR TITLE
fuchsia_remote_debug_protocol allows open port on remote device

### LINKED ISSUE #63968: fuchsia_remote_debug_protocol support port forwarding mechanisms that open a remote port
## Use case
Fuchsia is replacing ssh port forwarding in e2e tests where the test logic runs on a test host, and the Dart VM runs on a separate device under test (DUT). Port forwarding is used so that the test can access the Flutter debug ports that accept only local connections. The proposed replacement for ssh in this case is to run a separate program on the DUT that listens for external connections on some other port p and forward those requests to the Flutter debug port. In this scheme, to communicate with the debug port the test should call DUT:p instead of a localhost address as with ssh.

## Proposal
Currently the fuchsia_remote_debug_protocol package allows defining a custom port forwarding method to use to connect to a Flutter instance. However, it only allows for port forwarding mechanisms where the open port called from the test is some port on the test host (test accesses the debug port by calling localhost:x).  I propose expanding fuchsia_remote_debug_protocol to support cases where a forwarding mechanism opens a port on the DUT (test accesses debug port by calling DUT:x).

### PR DESCRIPTION
**Pull request by satsukiu**
Allows defining a port forwarding function for which the accessible
port is not on the host device. Examples include tunneling solutions
where a tunneling program on the same device as the Dart VM exposes
an open port through which it tunnels connections to the VM.

## Description

This change extends the functionality of the fuchsia_remote_debug_protocol package. Currently, the package allows forwarding ports from a host device to some other target device running a Flutter instance, which enables the host device to access the debug port exposed by the Flutter instance. 
This updates the definition of a port forwarding function to allow for cases where the host device accesses the debug port by calling a port on the target device rather than the local loopback of the host device.

## Related Issues

Resolves https://github.com/flutter/flutter/issues/63968

## Tests

I added the following tests:
* A copy of the existing end-to-end test that verifies that the correct URIs are used to connect to a remote VM. The new copy uses a remote IP address rather than local loopback.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jonahwilliams** [comment]
I do not have the context to review changes to this package. Unfortunately I am not aware if we have integration tests to exercise it within the flutter tree either

File: packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart
Comment (zanderso) [inline packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:408]: Does this comment need an update?

**Review by zanderso — COMMENTED** [review]
We don't really have any integration tests on Flutter's CI that will touch this code. Can you describe how you validated this change locally?

**Comment by satsukiu** [comment]
Thanks for taking a look!
I validated this by patching in the change to the Fuchsia tree and running some e2e tests that use this library. They're tests that run on a linux host, connect to a remote Fuchsia device, and drive flutter on the Fuchsia device.
There's one here that I ran - https://fuchsia.googlesource.com/fuchsia/+/refs/heads/master/src/tests/end_to_end/perf/test/flutter/scroll_flutter_benchmarks_test.dart#41
In addition there's a number of more complex internal tests that also use this library I validated this against.

I tried this both with and without my intended changes to the Fuchsia codebase (https://fuchsia-review.googlesource.com/c/fuchsia/+/421499), to verify that this isn't breaking and that it works for the case I need.

File: packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart
Comment (satsukiu) [reply packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart]: This comment is still accurate, but it's specific to the SSH port forwarder. Merged it with the comment in _SshPortForwarder.start.

**Review by zanderso — APPROVED** [review]
lgtm

/cc @awdavies who is the original author of these files and may want to take a look as well.

**Comment by goderbauer** [comment]
@zanderso @awdavies @satsukiu is this one ready to go in?

**Comment by satsukiu** [comment]
Thanks for checking, I'd like to merge this change unless awdavies@ has comments. I've reached out to him last week and today, could we wait another day before merging?

**Comment by zanderso** [comment]
@satsukiu It looks like you need to rebase past the unrelated "Windows tool_tests" failures to get the presubmit checks to turn green. Yes, we can give awdavies@ time to take a look.

**Comment by satsukiu** [comment]
All right, I haven't gotten a response from awdavies@ since - could we merge this when the build is healthy again? Thanks!

**Review by raggi — COMMENTED** [review]
LGTM

**Pull request merged**
